### ISSUE TITLE
tf.nn.batch_normalization returns incorrect gradients for scale (gamma): violates linearity w.r.t. upstream gradients (output_gradients) on TF 2.3.0 and 2.17.0 (silent correctness bug)

### ISSUE THREAD

**Issue by XiaoHuaOvo-tech** [tensorflow/tensorflow#107705]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

TF2.17.0

### Custom code

Yes

### OS platform and distribution

Linux Ubuntu 22/04

### Python version

3.10.19

### CUDA/cuDNN version

CUDA version: 12.3 cuDNN version: 8

### GPU model and memory

Tesla V100

### Current behavior?

Describe the bug
We observed a silent gradient correctness bug in TensorFlow where tf.nn.batch_normalization produces incorrect gradients w.r.t. the scale parameter (gamma). The gradient violates a fundamental invariant:
﻿
For a fixed forward graph, gradients must be linear in upstream gradients:

g_gamma(delta * G) == delta * g_gamma(G) for any scalar delta.

However, on TensorFlow 2.3.0 and 2.17.0, we consistently observe cases where:
﻿
g_gamma(G) == [0, 0, 0] (exact zeros), but
﻿
g_gamma(delta*G) becomes significantly non-zero,
﻿
so g_gamma(delta*G) != delta * g_gamma(G).
﻿
This is a silent correctness issue (no crash / no NaNs required), and BatchNorm is widely used, so incorrect gamma gradients can silently corrupt training updates and harm reproducibility and model quality.

Validation metric:
We report the Predoo-style vR metric, defined as max_i |A_i − B_i| / (atol + rtol × max(|A_i|,|B_i|)).
vR > 1 indicates that the deviation exceeds the absolute + relative tolerance.
p is the coverage ratio.
E is a violation of intensity
﻿
Versions tested
﻿
TensorFlow 2.17.0: reproduces

### Standalone code to reproduce the issue

```shell
import os
import numpy as np
import tensorflow as tf

os.environ.setdefault("TF_CPP_MIN_LOG_LEVEL", "1")

def compute_vR(A, B, atol=1e-5, rtol=1e-3):
    A_flat = np.asarray(A).flatten()
    B_flat = np.asarray(B).flatten()
    N = len(A_flat)
    diff = np.abs(A_flat - B_flat)
    scale = np.maximum(np.abs(A_flat), np.abs(B_flat))
    tolerance = atol + rtol * scale
    eps = np.finfo(np.float64).tiny
    r_i = diff / (tolerance + eps)
    v_R = np.max(r_i)
    violation = v_R > 1.0
    violation_mask = r_i > 1.0
    rho = np.sum(violation_mask) / N
    E_mean = np.mean(np.maximum(0, r_i - 1.0))
    worst_idx = np.argmax(r_i)
    worst_info = {
        'idx': int(worst_idx), 'r_i': float(r_i[worst_idx]),
        'diff': float(diff[worst_idx]), 'tolerance': float(tolerance[worst_idx]),
        'A_val': float(A_flat[worst_idx]), 'B_val': float(B_flat[worst_idx])
    }
    return v_R, violation, rho, E_mean, worst_info

def main():
    print("TF:", tf.__version__)
    print("Devices:", tf.config.list_physical_devices())

    np.random.seed(0)
    tf.random.set_seed(0)

    x = tf.random.normal((10, 32, 32, 3), dtype=tf.float32)
    gamma0 = tf.ones([3], dtype=tf.float32)
    beta0  = tf.zeros([3], dtype=tf.float32)
    eps = 1e-5

    G1 = tf.ones_like(x)
    delta = tf.constant(-2.097364, dtype=tf.float32)
    G2 = delta * G1

    with tf.GradientTape(persistent=True) as tape:
        x_v     = tf.Variable(x)
        gamma_v = tf.Variable(gamma0)
        beta_v  = tf.Variable(beta0)

        mean, var = tf.nn.moments(x_v, axes=[0,1,2], keepdims=False)
        out = tf.nn.batch_normalization(x_v, mean, var, beta_v, gamma_v, eps)

    # --- gGamma via output_gradients ---
    g1 = tape.gradient(out, gamma_v, output_gradients=G1)
    g2 = tape.gradient(out, gamma_v, output_gradients=G2)

    # --- analytical reference for GMR-2 ---
    inv = tf.math.rsqrt(var + eps)
    xhat = (x_v - mean) * inv
    g1_ref = tf.reduce_sum(G1 * xhat, axis=[0,1,2])

    del tape

    g1_np = g1.numpy()
    g2_np = g2.numpy()
    g1_ref_np = g1_ref.numpy()
    delta_np = float(delta.numpy())

    print("\nRaw values:")
    print("gGamma(G1)      :", g1_np)
    print("gGamma(G2)      :", g2_np)
    print("delta*gGamma(G1):", (delta_np * g1_np))
    print("Theory gGamma   :", g1_ref_np)

    # ---------------- GMR-1: linearity ----------------
    A = g2_np
    B = delta_np * g1_np
    vR, vio, rho, E, worst = compute_vR(A, B, atol=1e-5, rtol=1e-3)
    print("\n[GMR-1] Linearity: gGamma(delta*G) == delta*gGamma(G)")
    print(f"vR={vR:.6f}, violation={vio}, rho={rho:.2%}, E_mean={E:.6f}")
    print(f"worst: idx={worst['idx']} A={worst['A_val']:.6f} B={worst['B_val']:.6f} "
          f"diff={worst['diff']:.6f} tol={worst['tolerance']:.6f} r_i={worst['r_i']:.6f}")

    # ---------------- GMR-2: structure ----------------
    A = g1_np
    B = g1_ref_np
    vR, vio, rho, E, worst = compute_vR(A, B, atol=1e-5, rtol=1e-3)
    print("\n[GMR-2] Structure: gGamma == sum(G * xhat)")
    print(f"vR={vR:.6f}, violation={vio}, rho={rho:.2%}, E_mean={E:.6f}")
    print(f"worst: idx={worst['idx']} A={worst['A_val']:.6f} B={worst['B_val']:.6f} "
          f"diff={worst['diff']:.6f} tol={worst['tolerance']:.6f} r_i={worst['r_i']:.6f}")

if __name__ == "__main__":
    main()
```

### Relevant log output

```shell
TF2.17.0
Validation metric:
We report the Predoo-style vR metric, defined as max_i |A_i − B_i| / (atol + rtol × max(|A_i|,|B_i|)).
vR > 1 indicates that the deviation exceeds the absolute + relative tolerance.
In both GMR-1 and GMR-2, we observe vR ≈ 3–6 and ρ = 1.0 (100 % of elements violating),
confirming a systematic, reproducible gradient-computation defect rather than random numerical noise.
Raw values:
gGamma(G1)      : [0. 0. 0.]
gGamma(G2)      : [ 2.2699713e-05  3.0506477e-05 -1.3375095e-05]
delta*gGamma(G1): [-0. -0. -0.]
Theory gGamma   : [ 5.3405762e-05 -2.2888184e-05  5.7220459e-05]

[GMR-1] Linearity: gGamma(delta*G) == delta*gGamma(G)
vR=3.041370, violation=True, rho=100.00%, E_mean=1.213974
worst: idx=1 A=0.000031 B=-0.000000 diff=0.000031 tol=0.000010 r_i=3.041370

[GMR-2] Structure: gGamma == sum(G * xhat)
vR=5.689490, violation=True, rho=100.00%, E_mean=3.428429
worst: idx=2 A=0.000000 B=0.000057 diff=0.000057 tol=0.000010 r_i=5.689490
```

**Comment by nguyenvantam00** [comment]
I can confirm this issue reproduces on T**ensorFlow 2.17.0** and also on **2.3.0** using the provided script.

I additionally tested **TensorFlow 2.8.0**, and c**ould not reproduce the gradient linearity violation**
(gGamma(delta·G) == delta·gGamma(G) holds within tolerance), though this may need further confirmation across devices.

This suggests the bug may have been **introduced after the BatchNorm kernel / fused backward rewrite**, potentially in the gamma gradient reduction path.

I also forced **CPU-only execution** by setting
os.environ["CUDA_VISIBLE_DEVICES"] = "-1" before importing TensorFlow.
In this configuration, only CPU is visible (tf.config.list_physical_devices("GPU") == []),
and **the issue could not be reproduced**, which suggests the problem may be **specific to the CUDA fused BatchNorm backward kernel** rather than the mathematical formulation itself.

It would be helpful to confirm whether this is:

- "specific to the **fused BatchNorm backward CUDA kernel**, or"

- "backend-dependent (CUDA vs CPU / XLA),"

- "or related to a particular **reduction / accumulation path for gamma gradients.**"

<img width="2068" height="1566" alt="Image" src="https://github.com/user-attachments/assets/108007e2-fa56-42f0-8f07-cc84e91ab117" />

**Comment by nguyenvantam00** [comment]
I can confirm the issue reproduces on TensorFlow 2.17.0 in a CPU-only environment.

- TensorFlow version: 2.17.0
- Device: CPU only (no CUDA / no GPU detected)
- OS: Ubuntu 22.04
- Python: 3.10

The provided repro script runs without modification and shows the same
gamma gradient linearity violation.

This indicates the issue is not CUDA-specific and reproduces on the CPU path as well.

**Comment by Venkat6871** [comment]
Hi @XiaoHuaOvo-tech  ,
Thank you so much for filing this issue. We truly appreciate your support. At the moment, this request is not prioritized and mark it for community contributions. If you would like to contribute on a fix, please feel free to fork the repository and submit a pull request in accordance with the TensorFlow [CONTRIBUTION GUIDELINES](https://www.tensorflow.org/community/contribute/code) and link this issue in your PR to ensure the maintainers can easily review and collaborate.
To avoid duplicate work, please leave a comment here if you decide to take it on!

Thank you again for your interest and we look forward to seeing your contributions!

**Comment by nguyenvantam00** [comment]
thask you
Vào Thứ 4, 26 thg 8, 2026 lúc 20:28 Venkat6871 ***@***.***>
đã viết:

**Comment by vishwakt** [comment]
The reported quantity has a true mathematical value of exactly zero: with upstream gradient `G = ones`, the gamma gradient is `rsqrt(var+eps) * (sum(x) - N*mean)` per channel, and `mean` is by construction `sum(x)/N` over the same axes. All the values in the report (`0`, `~2e-5`, `~5e-5`) are float32 cancellation noise around that true zero; a 10*32*32 reduction has a noise floor around 1e-4, which is above the chosen `atol=1e-5`. Running the identical script in float64 collapses every deviation to ~1e-12, which is the signature of rounding rather than an incorrect gradient formula. Bitwise linearity `g(delta*G) == delta*g(G)` cannot hold in floating point for a non-power-of-two `delta`, and near a true zero the relative `vR` metric is unbounded by construction. So the exact zeros for `g_gamma(ones)` are the correct answer, and the comparison values are alternative rounding patterns of the same zero, not evidence of a defect.
